02-intermediate-ruby/inheritance-idioms.md

```diff
@@ -0,0 +1,255 @@
+# Inheritance Idioms
+
+## Learning Goals
+
+By the end of this lesson, students should be able to...
+
+- Define the following terms
+  - Abstract class
+  - Template method
+  - Polymorphism
+- Write code that extends an _abstract class_
+- Identify and implement _template methods_
+- Take advantage of _polymorphism_
+
+## Introduction
+
+There are many common patterns around how inheritance is used in real software projects that are worth studying explicitly. Today we will discuss three of these: _abstract classes_ and _template methods_, and _polymorphism_.
+
+### What is an Idiom?
+
+In the context of software, the word "idiom" means much the same thing as in regular English:
+
+> Idiom: A style or form of artistic expression that is characteristic of an individual, a period or movement, or a medium or instrument<br>
+> -- <cite>[Miriam Webster](https://www.merriam-webster.com/dictionary/idiom)</cite>
+
+In essence, a software idiom is a pattern or style of programming that is commonly used, but which isn't necessarily obvious from the syntax of the language.
+
+For an example of an idiom that doesn't involve inheritance, consider the `map` method. Originally most programming languages did not contain a method named `map` - if you wanted to do that work, you would have to write it yourself:
+
+```ruby
+# Map without map
+original = [1, 2, 3]
+result = []
+
+original.each do |item|
+  result << 2 * item
+end
+```
+
+Over the years programmers repeated this pattern over and over, to the point that "mapping" was a thing you could do without thinking about it. If you were to say something like "map this to that", people would know what you were talking about, even though there was nothing in the programming language called `map`. At this point, `map` had become an idiom.
+
+If an idiom becomes common enough, language designers will take note and include formal support for that pattern in their languages. The language LISP included something similar to `map` as far back as 1959, but most mainstream languages did not support `map` officially until between 2005 and 2010.
+
+The two idioms we will discuss today do not have language support in Ruby (though they do in other languages), but they are no less important for that.
+
+### Abstract Classes
+
+Again, let us begin with a definition:
+
+> Abstract: disassociated from any specific instance<br>
+> -- <cite>[Miriam Webster](https://www.merriam-webster.com/dictionary/abstract)</cite>
+
+**An _abstract class_ is a class that is never instantiated**. Instead, it serves as the parent class for some number of child classes, which are instantiated.
+
+**Question:** What abstract classes have we seen so far?
+
+<details>
+<summary>Click for the answer</summary>
+Object, CsvRecord from the project
+</details>
+
```

Comment:
Somewhere in this section, it could be useful to call out: Why are abstract classes used? Maybe an example could help.
The abstract class provides a contract for all the derived sub-classes to follow. Without the contract being satisfied, the derived sub-classes cannot be instantiated.

_Aside_: I don't talk about abstract classes and templates. While talking about inheritance and polymorphism, I use the example of Car class that a Tesla class gets extended from. See slide 22 on https://drive.google.com/drive/folders/0B__DV26QHsH4dXJTRXlJc1oydGM